plugins/modules/ndo_ptp_policy_profiles.py

```diff
@@ -0,0 +1,275 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2024, Shreyas Srish (@shrsr) <ssrish@cisco.com>
+
+# GNU General Public License v3.0+ (see LICENSE or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {"metadata_version": "1.1", "status": ["preview"], "supported_by": "community"}
+
+DOCUMENTATION = r"""
+---
+module: ndo_ptp_policy_profiles
+short_description: Manage PTP Policy Profiles on Cisco Nexus Dashboard Orchestrator (NDO).
+description:
+- Manage PTP Policy Profiles on Cisco Nexus Dashboard Orchestrator (NDO).
+- This module is only supported on ND v3.1 (NDO v4.3) and later.
+author:
+- Your Name (@yourusername)
+options:
+  template:
+    description:
+    - The name of the template.
+    - The template must be a fabric policy template.
+    type: str
+    required: true
+  ptp_policy_profile_name:
+    description:
+    - The name of the PTP Policy profile.
+    type: str
+    aliases: [ name ]
+  ptp_policy_profile_uuid:
+    description:
+    - The UUID of the PTP Policy profile.
+    type: str
+    aliases: [ uuid ]
+  delay_interval:
+    description:
+    - Delay interval of the PTP Policy profile.
+    - The value should be between -4 to 4.
+    - This is required when O(state=present).
```

Comment:
Are attributes like `delay_interval` which are described as `required when O(state=present)` needed for creation only or everytime the user manipulates a PTP Policy Profile (creation + update)?

If creation only, does the API error out on its own? if not, we should create flags to check if the attributes exist but only for creation steps.

And the same thing applies for ndo_ptp_policy?

Reply:
When it comes to the module ndo_ptp_policy and we take the two below configs as example where fabric_delay_interval is missing from the second config.
```
   - name: 1st config
      cisco.mso.ndo_ptp_policy: 
        <<: *mso_info
        template: ansible_fabric_policy_template
        ptp_policy: ansible_test_ptp_policy
        fabric_sync_interval: -3
        global_domain: 0
        fabric_delay_interval: -2
        global_priority1: 255
        global_priority2: 255
        fabric_announce_timeout: 3
        fabric_announce_interval: 1
        fabric_profile_template: aes67-2015
        state: present
        
   - name: 2nd config
      cisco.mso.ndo_ptp_policy: 
        <<: *mso_info
        template: ansible_fabric_policy_template
        ptp_policy: ansible_test_ptp_policy
        fabric_sync_interval: -3
        global_domain: 0
        global_priority1: 255
        global_priority2: 255
        fabric_announce_timeout: 3
        fabric_announce_interval: 1
        fabric_profile_template: aes67-2015
        state: present
 ```
Technically the 2nd config should show as "no change" but it does for this module because the payload has an extra key "global" for a few sub-attributes and their values as seen below:
```
mso_values = {
            "name": ptp_policy,
            "description": description,
            "global": {
                "adminState": admin_state,
                "fabSyncIntvl": fabric_sync_interval,
                "globalDomain": global_domain,
                "fabDelayIntvl": fabric_delay_interval,
                "prio1": global_priority1,
                "prio2": global_priority2,
                "fabAnnounceTimeout": fabric_announce_timeout,
                "fabAnnounceIntvl": fabric_announce_interval,
                "fabProfileTemplate": fabric_profile_template,
            },
        }
 ```
When we don't include fabric_delay_interval above, the new append_update_ops_data() actually detects a change in the entire value of global and not fabDelayIntvl specifically. As a result a null value is passed to fabDelayIntvl which NDO consumes as "0" and this is actually a valid value for fabDelayIntvl.

For now I added this to prevent a change which enforces the user to actually specify it every time.
```
[
                "state",
                "present",
                [
                    "global_priority1",
                    "global_priority2",
                    "global_domain",
                    "fabric_announce_timeout",
                    "fabric_announce_interval",
                    "fabric_sync_interval",
                    "fabric_delay_interval",
                ],
            ],
```
What we can do is not use append_update_ops_data() at all for this particular module and have individual checks like before.